src/ldp/permissions/SparqlPatchPermissionsExtractor.ts

```diff
@@ -0,0 +1,58 @@
+import { Algebra } from 'sparqlalgebrajs';
+import { Operation } from '../operations/Operation';
+import { PermissionSet } from './PermissionSet';
+import { PermissionsExtractor } from './PermissionsExtractor';
+import { Representation } from '../representation/Representation';
+import { SparqlUpdatePatch } from '../http/SparqlUpdatePatch';
+import { UnsupportedHttpError } from '../../util/errors/UnsupportedHttpError';
+
+/**
+ * Generates permissions for a SPARQL DELETE/INSERT PATCH.
+ * Updates with only an INSERT can be done with just append permissions,
+ * while DELETEs require write permissions as well.
+ */
+export class SparqlPatchPermissionsExtractor extends PermissionsExtractor {
+  public async canHandle(input: Operation): Promise<void> {
+    if (input.method !== 'PATCH') {
+      throw new UnsupportedHttpError('Only PATCH operations are supported.');
+    }
+    if (!input.body) {
+      throw new UnsupportedHttpError('PATCH body is required to determine permissions.');
+    }
+    if (!this.isSparql(input.body)) {
+      throw new UnsupportedHttpError('Only SPARQL update PATCHes are supported.');
+    }
+    if (!this.isDeleteInsert(input.body.algebra)) {
+      throw new UnsupportedHttpError('Only DELETE/INSERT SPARQL update operations are supported.');
+    }
+  }
+
+  public async handle(input: Operation): Promise<PermissionSet> {
+    if (!input.body || !this.isSparql(input.body) || !this.isDeleteInsert(input.body.algebra)) {
+      throw new UnsupportedHttpError('A SPARQL DELETE/INSERT body is required.');
+    }
+    const result = {
+      read: false,
+      append: this.needsAppend(input.body.algebra),
+      write: this.needsWrite(input.body.algebra),
+    };
+    result.append = result.append || result.write;
+    return result;
+  }
+
+  private isSparql(data: Representation): data is SparqlUpdatePatch {
+    return Boolean((data as SparqlUpdatePatch).algebra);
+  }
+
+  private isDeleteInsert(op: Algebra.Operation): op is Algebra.DeleteInsert {
+    return op.type === Algebra.types.DELETE_INSERT;
+  }
+
+  private needsAppend(update: Algebra.DeleteInsert): boolean {
+    return Boolean(update.insert && update.insert.length > 0);
```

Comment:
```suggestion
    return update.insert?.length > 0;
```

Reply:
This actually does not work. The `?` operator potentially changes the entire block it is part of to `undefined` if the left part is `undefined`, but this can not be compared to 0. This can be seen in the javascript code that would be generated from the suggestion above.
```javascript
        var _a;
        return ((_a = update.insert) === null || _a === void 0 ? void 0 : _a.length) > 0;
```
Took me some time to wrap my head around initially, but the `?` operator can only be used in instances where you're checking if something exists or not.

Reply:
What a shame.

Reply:
Hold on 🙂

`undefined > 0` is false in JS, which is what we want?

Reply:
Yea but TypeScript doesn't want you to start comparing undefined with numbers ;). Due to the `?` operator the type of the left part is `undefined | number`.